### PR TITLE
Overhaul of horizontal (interpolatory) remap

### PR DESCRIPTION
**Pull request by bartgol**
This PR reorganizes a bit the implementation of horizontal interpolatory remapping. A summary of changes/additions

- Add a refining remapper. Analogue to CoarseningRemapper, this remapper goes in the opposite direction. As a consequence, to minimize amount of data communicated, we do the mat-vec after the MPI communication.
- The refining remapper comes in two flavor: classic two-sided MPI, and one-sided MPI. The latter has some limitation (e.g., no device pointers), but I wanted to have a good apples-to-apples comparison test for MPI's RMA approach (all benchmarks I can find online do different things for RMA and P2P tests, so it's hard to gauge the costs/benefits). 
- Add a HorizInterpRemapperBase class, where we can put common code shared by both RefiningRemappersXYZ as well as CoarseningRemapper. For now, I only added impl that retrieves the triplets from the map file. I am planning to move mat-vec implementation there as well, to avoid having 2-3 copies of it.
- Add some utilities to gather/scatter grid-related data. Right now, they are only used by the P2P remapper, as well as by HorizInterpRemapperBase. I am planning to extent their use, and remove some hard-to-read functions in CoarseningRemapper (namely, `get_pids_for_recv` and `recv_gids_from_pids`).
- Added unit tests for the refining remappers, analogue to those for the coarsening remapper. In fact, their map file is basically the transpose of that in the coarsening remapper tests.

Todo:
- use more of the import-export utilities in the remappers
- move mat-vec in HorizInterpRemapperBase
- consolidate remappers tests in a single unit test (they have the same setup/checks)
- make rma remapper compilation optional (I don't want to risk these tests to cause failures on some machines, due to RMA shenanigans).
- check if `horizontal_remap_utility.*pp` is still needed, or if we can just use the remapper. Not a priority, we can leave this for SPA's sake, but long run it may be confusing to have two things that do roughly the same thing.

@mahf708 If you want, you can start using the RefiningRemapperP2P in your work on nudging. The interface is written in stone (it's the AbstractRemapper interface), so you can rely on that. And since the p2p tests pass (on my machine, we'll see about GPU), you can already start working without fear of compilation errors and/or crashes (again, at least on CPU, I'll see what the AT says about GPU).

### CODE REVIEW

**Comment by AaronDonahue** [comment]
@bartgol I haven't looked through all your code changes, so apologies if you already do this in this PR.  But reacting to your comment about SPA... I think long-term we do want to have SPA just use whatever internal remapper routines we are already using elsewhere in the code.

**Comment by bartgol** [comment]
@tcclevenger @AaronDonahue I pinged you both, since this PR is a bit large, so two sets of eyes may help.

I disabled the one-sided MPI version of the refining remapper by default. I will use it to do some experiments on my own, but for now we should probably stick with P2P MPI. Feel free to review the RMA one as well though, if you want.

The RefiningRemapper is not used anywhere (except unit tests). I think that's the best thing to do: merge the PR, and then make another PR to add it to ML or SPA processes (when the latter is done, remove the host-only horiz interpolation files).

File: components/eamxx/src/share/tests/refining_remapper_tests.cpp [hunk 1]
```diff
@@ -0,0 +1,467 @@
...
+          REQUIRE (weights_h(j)==1);
+        } else {
+          REQUIRE (weights_h(j)==0.5);
+        }
+      }
+    }
+  }
+};
+
+Field create_field (const std::string& name, const LayoutType lt, const AbstractGrid& grid)
```
Comment (AaronDonahue) [inline components/eamxx/src/share/tests/refining_remapper_tests.cpp]: None of these fields are packed.  Would it make sense just to include at least one packed field to have more coverage?

**Comment by tcclevenger** [comment]
@bartgol I've started taking a look. Will continue tomorrow.

**Review by AaronDonahue — COMMENTED** [review]
Looks good to me.  Quite the PR, well done Luca!  I had two comments about packsize in the remapping tests.  Feel free to push back.

**Comment by AaronDonahue** [comment]
@bartgol the PR description still describes this PR as being "pratically WIP..."  If it's ready to review you can delete that phrase.

File: components/eamxx/src/share/tests/refining_remapper_rma_tests.cpp [hunk 2]
```diff
@@ -0,0 +1,477 @@
...
+          REQUIRE (weights_h(j)==1);
+        } else {
+          REQUIRE (weights_h(j)==0.5);
+        }
+      }
+    }
+  }
+};
+
+Field create_field (const std::string& name, const LayoutType lt, const AbstractGrid& grid)
```
[the remark this answers is not in the document]
Comment (bartgol) [reply components/eamxx/src/share/tests/refining_remapper_rma_tests.cpp:73]: Sure, I can make nlevs=SCREAM_PACK_SIZE, and always pack 3d fields. That should yield perfect packing for midpoints fields and padded packing for interface fields.

[the remark this answers is not in the document]
Comment (bartgol) [reply components/eamxx/src/share/tests/refining_remapper_rma_tests.cpp:73]: That also reminds me that there is a bug for padded fields, which I need to fix.

**Review by AaronDonahue — APPROVED** [review]
thanks for making those changes.

**Comment by AaronDonahue** [comment]
@bartgol do you think this is ready to go?  Could we add the AUTOMERGE label?  Do we still want @tcclevenger to take a look?

**Comment by bartgol** [comment]
He said he was going to, so I wanted to give him time.

**Comment by tcclevenger** [comment]
@bartgol @AaronDonahue I was able to look over most of it, and it looks good. I'll add the automerge. 

Just curious, were you able to test the performance difference P2P vs RMA?

**Review by tcclevenger — APPROVED** [review]

**Comment by bartgol** [comment]
> Just curious, were you able to test the performance difference P2P vs RMA?

Not yet. I tested the RMA refiner against the coarsener, and it was lower, but the pattern is different. Also, I talked with some kokkos dev about using [kokkos remote spaces](https://github.com/kokkos/kokkos-remote-spaces), which according to the dev has comparable (or better) perf than P2P for a mat-vec kind of communication. So rather than spend time with RMA (which does not support dev pointers anyways), I may wait until I have time to experiment with that, and then to a real comparison.

**Comment by AaronDonahue** [comment]
@bartgol , is this now ready?

**Comment by bartgol** [comment]
Not yet. Testing for timing is currently ongoing on chrys.

**Comment by bartgol** [comment]
Ok, timers seem normal. I'll re-run AT, since the PR has been marked as stale, then we can merge.

**Pull request merged**